### PR TITLE
Add generic message pump and let StorageQueueMessagePump use it

### PR DESCRIPTION
**Pull request by NiklasArbin**
It's pretty hard to write a pump without bugs and concurrency issues, let's provide a default implementation

### CODE REVIEW

**Comment by NiklasArbin** [comment]
And just to prove my point I introduced a concurrency bug

File: knightbus/src/KnightBus.Core/GenericMessagePump.cs [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                if (!await PumpAsync<TMessage>(action, cancellationToken).ConfigureAwait(false))
+                    await DelayPolling().ConfigureAwait(false);
+            }
+        }, CancellationToken.None);
+        return Task.CompletedTask;
+    }
+
+    public async Task<bool> PumpAsync<TMessage>(Func<TInternalRepresentation, CancellationToken, Task> action, CancellationToken cancellationToken)
```
Comment (viktor-h) [inline knightbus/src/KnightBus.Core/GenericMessagePump.cs:53]: can be internal och private accessibility level

**Review by viktor-h — APPROVED** [review]
Tested on postgres transport and it seems to work!

File: knightbus/src/KnightBus.Core/GenericMessagePump.cs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    protected virtual async Task DelayPolling()
+    {
+        try
+        {
+            await Task.Delay(PollingDelay, _pumpDelayCancellationTokenSource.Token).ConfigureAwait(false);
+        }
+        catch (TaskCanceledException)
+        {
+            //reset the delay
+            _pumpDelayCancellationTokenSource = new CancellationTokenSource();
```
Comment (andrevirdarson) [inline knightbus/src/KnightBus.Core/GenericMessagePump.cs:194]: Should we add a comment to why? I guess it is when the timeout token is triggered

**Review by andrevirdarson — APPROVED** [review]
Very nice! 🌟

File: knightbus/src/KnightBus.Core/GenericMessagePump.cs (hunk 1 above)
Comment (NiklasArbin) [reply knightbus/src/KnightBus.Core/GenericMessagePump.cs:53]: Hard to test if private

**Pull request merged**